mailmap: add support for `.mailmap` files

emilazy opened the pull request:
Explanation, motivation, backstory, and design decisions are in the commit message, and I’ve included comprehensive tests and hopefully‐adequate documentation.

Pulling in the people who actively discussed this on Discord for review, hope that’s okay: @nasamuffin @arxanas @PhilipMetzger

Some things I’m unsure about and could use feedback on:

* I’m happy to hear bikeshedding about the DSL interfaces here, especially the `*_raw` names (`*_unmapped`?).

* There is a breaking change to the Rust `.author()` and `.committer()` methods on `Commit`; I assume this is okay due to the current level of general churn, but I thought I should point it out. I plan to look into opening a PR to make gg do the right thing here.

* ~~`get_wc_commit_mailmap` isn’t exactly my favourite function name in the world, and it feels a little gross to have commit/tree‐related plumbing in the `mailmap` module (e.g., `jj_lib::gitignore` just has a pure model). Any suggestions for better places this could go, better names it could have, types it could hang off as a method of, etc. are welcome.~~

* ~~I added `pub(crate)` to `jj_lib::git_backend: signature_{from,to}_git`, as they’re used by the `mailmap` module to interface with `gix-mailmap`. Since this isn’t a feature specific to the Git backend, it feels wrong to have that dependency, but I’m not sure where would be good to move them to. A new `jj_lib::git_util` module, perhaps?~~

* ~~I don’t love the `jj_cmd_ok_as` thing in the CLI test, but it seemed like too much work to refactor that interface right now.~~

* I don’t really know how conflicts in the `.mailmap` file should be handled. Ideally the entries not involved in a conflict should still be respected, but I’m not even sure how `materialize_tree_value` behaves here.

* ~~Should I split this into multiple commits? A lot of the changes are fairly entangled, but I could probably tease one or two more commits out of this.~~

* Should there be more permanent documentation of the `.mailmap` support, and if so, where?

<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the branch rather than adding commits on top. Use force-push when
pushing the updated branch (`jj git push` does that automatically when you
rewrite a branch). Merge the PR at will once it's been approved. See
https://github.com/martinvonz/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

emilazy commented:
Looks like these ended up being immediately relevant to https://github.com/martinvonz/jj/actions/runs/9644975580/job/26598226134?pr=3951:

> That said, this is not exclusive to the Git backend. The `.mailmap` name and format is perfectly generic, already shared between Git and Mercurial, and applies to all systems that bake names and emails into commits, including the current local backend. The code uses Gitoxide, but only as a convenient implementation of the file format; in a hypothetical world where the Git backend was removed without Jujutsu changing its notion of commit signatures, `gix-mailmap` could be used standalone, or replaced with a bespoke implementation.

> I added `pub(crate)` to `jj_lib::git_backend: signature_{from,to}_git`, as they’re used by the `mailmap` module to interface with `gix-mailmap`. Since this isn’t a feature specific to the Git backend, it feels wrong to have that dependency, but I’m not sure where would be good to move them to. A new `jj_lib::git_util` module, perhaps?

I can add a separate direct `gix_mailmap` dependency, rather than adding it to the main `gix` dependency, to signal that it’s not being depended on as part of the Git backend and make the check happy. Suggestions for where to move the signature conversion functions would be appreciated though!

PhilipMetzger commented on lib/src/mailmap.rs:
```diff
@@ -0,0 +1,93 @@
+
+/// Reads and parses the `.mailmap` file from the working copy commit of the
+/// specified workspace.
+pub fn get_wc_commit_mailmap(repo: &dyn Repo, workspace_id: &WorkspaceId) -> Mailmap {
```
Nit: I think this should just return a `Option<Mailmap>` as it isn't relevant to all repos.

emilazy replied:
An absent `.mailmap` is the same as an empty one, and is represented under the hood by an empty `Vec` (which does no allocation), so I’m not sure this is desirable. I had it like that previously, but every caller just did `.unwrap_or_default()`. Anything that cares more specifically about the presence and structure of the `.mailmap` file in the repository will have to poke at the tree APIs directly anyway.

PhilipMetzger replied:
while having a `Option` makes this immediately clear, having it in the doc comment should also be enough.

emilazy replied:
Good idea, thank you! Hopefully what I’ve pushed just now covers it. (If you can think of a better name or location for this function I’m all ears, too…)

PhilipMetzger replied:
Basically, I'd write an empty `.mailmap` is as cheap as an empty `String` or so.

emilazy replied:
I’ve added a comment to the type; let me know what you think. Looking up the `.mailmap` probably involves allocations sometimes (e.g. to interface with the backend or to materialize the tree), but I’ve conveyed that the actual empty `Mailmap` itself is cheap.

PhilipMetzger replied:
I think your comment is good enough, as it makes it immediately clear why we pass around a `MailMap::default()` everywhere. It also does not imply that a lookup is cheap.

PhilipMetzger commented on lib/src/mailmap.rs:
```diff
@@ -0,0 +1,93 @@
+// Copyright 2021 The Jujutsu Authors
```
Nit: 2024

emilazy replied:
Fixed, thank you!

PhilipMetzger commented on lib/src/revset.rs:
```diff
@@ -2000,12 +2023,14 @@ impl<'a> RevsetParseContext<'a> {
         user_email: String,
         extensions: &'a RevsetExtensions,
         workspace: Option<RevsetWorkspaceContext<'a>>,
+        mailmap: Rc<Mailmap>,
```
Nit: This also should be a `Option`

PhilipMetzger reviewed (commented):
First round, I'm not totally onboard for the `*_raw()` templates but let's wait for other peoples opinion. And to fix the build failure, wrap the necessary stuff in `#[cfg(git)]` and make the dependency rely on it.

emilazy commented:
I factored out the conversion functions into a new private `jj_lib::signature_util` module (bikeshedding on the name etc. welcome), and hopefully the tests will run now. As I explained in the commit message this isn’t specific to the Git backend (`gix-mailmap` is just an implementation detail), so nothing there is conditional.

emilazy commented:
For what it’s worth, the best alternative DSL API I came up with was `commit.{author,committer}([canonical=true])` and `{author,committer}{pattern, [canonical=true])`, but that required adding support for keyword arguments to the template DSL and didn’t seem obviously better than what I ended up with currently.

emilazy commented:
(Gah, messed up when splitting up commits; force‐pushed again.)

yuja commented on lib/src/signature_util.rs:
```diff
@@ -0,0 +1,118 @@
+
+use crate::backend::{MillisSinceEpoch, Signature, Timestamp};
+
+const EMPTY_STRING_PLACEHOLDER: &str = "JJ_EMPTY_STRING";
```
iirc, this placeholder value is a workaround specific to the git backend, which can't store an empty string. It shouldn't apply to the default (and maybe Google) backends.

emilazy replied:
Right, that makes sense. In this case it would only be relevant if you did something very weird like put `Nobody At All <JJ_EMPTY_STRING>` in the `.mailmap` file. Maybe I could add a special case to the `.mailmap` code to handle that? Or this part of the conversion could be factored out to be directly in the Git backend instead.

emilazy replied:
I could also just revert this code migration and hand‐code the conversion in the `mailmap` module. Without timestamp conversion or empty name handling it’s very brief so the remaining duplication would be fine, and it’s less code churn.

OTOH that would make the results inconsistent with Git when you are using the Git backend and try to put the empty user in `.mailmap`, but… well… probably don’t do that?

yuja replied:
I think our `Mailmap::resolve()` can just return early if the signature is empty. It's invalid in Git, and so empty name in mailmap would also be invalid.

Alternatively, maybe we can just reuse the parser part and do lookup by ourselves?

emilazy replied:
Returning early sounds reasonable to me. It will result in inconsistency with Git if anyone puts `<JJ_EMPTY_STRING>` maps in their `.mailmap`, but I don’t think that divergence matters considering how pathological a case that is.

yuja commented on lib/src/commit.rs:
```diff
@@ -145,10 +146,18 @@ impl Commit {
 
+    pub fn author(&self, mailmap: &Mailmap) -> Signature {
+        mailmap.resolve(&self.data.author)
+    }
```
Is the mailmap thingy a backend or frontend/UI matter? (I don't follow the previous discussion, so I don't know the conclusion, sorry.)

If it's backend stuff, I think the raw signature can be (forcibly) translated by `GitBackend` (or `Store`) layer. There might be a global flag to temporary turn off the translation.

If it's frontend stuff, the translation should be made explicitly by templater, and I think the revset frontend can do reverse translation (so the revset engine impls don't have to deal with the mailmap.)

emilazy replied:
It’s a frontend matter. You want canonical signatures to be used by default in the vast majority of contexts (hence exposing it as the “default” way to get this information out of a commit as a nudge), but there are use‐cases for inspecting and querying the raw information in the commits. Git offers a `--no-mailmap` option to do this, but offering it as separate DSL functions seemed more orthogonal and fitting with Jujutsu’s general interfaces. The translation also depends on the current working copy commit (to use the current `.mailmap` file), so it’s not trivial to do in a global way without plumbing through a lot more state.

yuja replied:
> It’s a frontend matter.

Thanks for clarification. In that case, I wouldn't add `Commit::author(mailmap)` functions, and just let templater/revset modules do translation.

> Git offers a `--no-mailmap` option to do this, but offering it as separate DSL functions seemed more orthogonal and fitting with Jujutsu’s general interfaces.

I'm not sure about the UI. I personally like opt-in interface (something like `author.mailmap().email()`, `author(mailmap:pattern)`), but that wouldn't be major opinion. If mailmap is enabled implicitly by default, and if it's unlikely to show both raw and translated signatures within the same session, adding `--no-mailmap` flag/config might be better than exposing `_raw()` DSL functions.

emilazy replied:
I was thinking about tooling using the library that wants to display metadata about a commit without using the templating language. If it’s entirely the caller’s responsibility to realize they should be handling the `.mailmap`, and there’s nothing in the API nudging them towards that or even just pointing out that they have to manually do it, I fear it would result in what @nasamuffin described as the whack‐a‐mole experience that Git has had to deal with of ensuring that everything that wants to render a name/email maps it correctly. My hope is that by making it explicit in the interface when you should expect to need the `.mailmap` information, and clear what you’re doing when you explicitly request the “wrong” results, this could be avoided.

Of course, I guess the existing `jj` command‐line tool always uses templates to display this kind of thing. But it seems like we can’t really count on that for external tools, and it’s surprising that to get the correct canonical author information as reported by `jj` from the library you’d need to layer on additional non‐obvious APIs? A doc comment might help here, but is still weak against autocompleting the obvious function compared to getting the machine to tell you what you need and what your alternatives are.

My concern with making it opt‐in in templating functions is the same – customizing your template and putting `author()` in and getting an outdated representation of the commit author that disagrees with what Git tooling will give you seems like a foot‐gun. Adding an `--ignore-mailmap` option to every command that queries revsets or formats commits would be possible, but it seems like a lot of additional CLI surface churn and state plumbing compared to just exposing functions for that? If people really don’t like the additional functions, maybe the `*_raw` variants should just be removed and people can use lower‐level tools to poke at the raw object data if they want (or even just temporarily delete the `.mailmap` file).

martinvonz replied:
I agree with Yuya that it should be kept in the templater. A quick way to check how much of a problem it will be in practice is to find where we call .author() and .committer() today.

yuja replied:
> I was thinking about tooling using the library that wants to display metadata about a commit without using the templating language. If it’s entirely the caller’s responsibility to realize they should be handling the `.mailmap`, and there’s nothing in the API nudging them towards that or even just pointing out that they have to manually do it, I fear it would result in what @nasamuffin described as the whack‐a‐mole experience that Git has had to deal with of ensuring that everything that wants to render a name/email maps it correctly. My hope is that by making it explicit in the interface when you should expect to need the `.mailmap` information, and clear what you’re doing when you explicitly request the “wrong” results, this could be avoided.

I see your point. I think it's up to how much guarantee (and Git compatibility) we're going to provide. (I personally don't care about mailmap-ed identity, so I wouldn't mind if jj output is diverged from git.)

> Adding an `--ignore-mailmap` option to every command that queries revsets or formats commits would be possible, but it seems like a lot of additional CLI surface churn and state plumbing compared to just exposing functions for that?

I believe it's actually easier to implement. The resulting code would be something like
```rust
let mailmap = if global_args.ignore_mailmap {
    vec![]
} else {
    laod_mailmap(repo.get_commit(wc))?
};
...
RevsetContext {
   ...
   mailmap,
}
```
That said, I'm not sure about UI/UX things.

emilazy replied:
I think it’s worth noting that consistent `.mailmap` support throughout tooling is a lot more important to some people than others. There are reasons for people to change name, like divorce or gender transition, where tooling unnecessarily surfacing their previous name can be both be upsetting to them and potentially confusing/misleading to others. It’s also important to people who want to accurately find commits by people who have changed their details, and there’s the case of people changing email due to things like loss of control over the account, where it would be a bad idea for people or automated systems to contact the old email because of its presence in a commit.

So, while I understand that to many people this feature will just be a triviality that they may not bother to use for their own information and that they feel they could take or leave for rendering, I think that the downside of not doing the right thing by default is a lot bigger than it might seem. Of course, we can’t solve this perfectly with the current data model, which bakes this data into commits permanently (hence this not resolving #2957), but I think it would be *really* unfortunate to do worse than Git here, and not just for reasons of compatibility for the sake of compatibility.

As far as the code structure goes – what about not exposing the mailmapping methods on `Commit`, but keeping the renames of the Rust API methods to `*_raw` (or `*_unmapped`?) and adding a doc comment explaining that the signature should be canonicalized before display or processing in most cases? That way the name will still signal to external users that there’s subtleties involved here, and hopefully make them inclined to check the documentation. The `jj` CLI doesn’t have many relevant calls here, but that’s because it extensively uses the templating language; I don’t think we can assume that every external tool that wants to get human‐suitable metadata out of commits will realize that they need to be indirectly going via the DSL, especially if the template function name stays the same as the Rust method name despite having different behaviour.

A global `--ignore-mailmap` option sounds okay to me, and I can implement it. It would mean that you can’t use the templating language to do things like compare raw signatures to what they’re getting mapped to, but that might not be a big deal.

martinvonz replied:
True, we want to make it easy for alternative UIs and such to get the mailmapped data, so I agree with you now that it should go in the lib crate. Thanks again for working on it. I'll try to find time to review it today.

emilazy commented:
Okay, I’ve dropped the canonicalizing functions from `Commit` and added corresponding convenience helpers on the `Mailmap` type instead, along with explanatory documentation on the non‐canonicalizing `Commit` methods, and pointers for obtaining a `Mailmap`. That seems better‐factored and should probably be sufficient to nudge tooling developers onto the right path. Thank you both for the feedback; hopefully this is better from a code organization POV.

I also separated out the signature conversion, and renamed `get_wc_commit_mailmap` to `get_current_mailmap`, which I dislike a lot less. It turns out that Git actually *does* support empty email addresses, and that real repositories (e.g. Nixpkgs) have such commits in them; it would be a practical use of `.mailmap` to correct these, e.g. `<real@email.example> Author Name <>`, so rather than bailing early on empty signatures I just convert them to Gitoxide’s types as‐is. It’s fewer Git‐specific details in backend‐agnostic code, less churn in the rest of the codebase, and overall better behaviour, so it turned out nicely.

I’ve currently left the `{author,committer}{,_raw}` revset and template functions rather than moving them to a global option. Git has e.g. `%an`/`%aN` pretty format pairs which are analogous to this; I assume they only have the option because it historically used to be opt‐in and they don’t have a real query language. It feels more flexible and declarative to encode the intent in the function names rather than having their semantics depend on more global state. But if people feel strongly about just having a global option to disable it I’m fine implementing that instead.

nasamuffin reviewed (commented):
Thanks for leaving the mechanical bits in a standalone commit.

GitHub won't let me comment on the commit message :) but in my opinion, you could say a little more - that the mailmap will introduce a concept of pre- and post-mapped identities. Right now, it's implied here.

Are committer and author the only types of identities that are here? What happens to, say, commit message footers like Signed-off-by: (which are mailmapped in Git, IIRC)? Is that handled differently?

emilazy commented:
Thank you for taking a look!

> GitHub won't let me comment on the commit message :) but in my opinion, you could say a little more - that the mailmap will introduce a concept of pre- and post-mapped identities. Right now, it's implied here.

That’s not something I hear often :)

I think this is a good idea and I’ve added another few paragraphs (after my editor ate the first draft…); I’ve called them *raw signatures* and *canonical signatures*, as that’s the best terminology I can think of currently and it matches the code and comments. I’ve also tweaked a few doc comments to allude to these concept more explicitly, although there isn’t an in‐depth exposition in the code currently. Possibly this should be recorded more permanently in developer documentation somewhere but hopefully that can be left to a follow‐up PR – the design docs section, perhaps? Although I wouldn’t exactly say that this is a stellar piece of de novo design to flaunt, just a pragmatic stop‐gap for arguably‐regrettable decisions the majority of DVCSes have made up to this point…

> Are committer and author the only types of identities that are here? What happens to, say, commit message footers like Signed-off-by: (which are mailmapped in Git, IIRC)? Is that handled differently?

As far as I can tell authors and committers are the only places Jujutsu gets and processes signatures from currently, and there is no special support for trailers. It seems like `git shortlog --group=…` canonicalizes identity information in trailers, but `git log` doesn’t on display; a strange state of affairs that speaks to the unfortunate whack‐a‐mole you talked about and that I’m trying to avoid repeating. I guess it makes some sense with the nature of trailers as an ad‐hoc interpretation layer on top of unstructured commit messages, which I also think is regrettable. Hopefully Jujutsu can at least do better there, even if it has to lower whatever additional metadata it ends up recording down to trailers for now for Git compatibility! I’ve discussed this in the commit message too as another example of where the raw–canonical signature distinction and guidelines for which to use would be relevant.

martinvonz commented on lib/Cargo.toml:
```diff
@@ -44,6 +44,9 @@ either = { workspace = true }
 futures = { workspace = true }
 git2 = { workspace = true, optional = true }
 gix = { workspace = true, optional = true }
+gix-actor = { workspace = true }
+gix-date = { workspace = true }
```
nit: do we need to declare these two? gix-mailmap seems to depend on them, so it should be enough to declare only that dependency, no?

yuja replied:
Maybe feature flag of `gix` can be set conditionally instead?

emilazy replied:
We need `gix_actor::SignatureRef` in `jj_lib::mailmap`, so that needs to be an explicit dependency. `gix-date` was formerly required but no longer is; I’ve removed it, thank you!

The `gix` crate has [a lot of dependencies](https://github.com/Byron/gitoxide/blob/8d89b865c84d1fb153d93343d1ce4e1d64e53541/gix/Cargo.toml#L293) even with all features disabled. They might fall out during linking, but I figured that it would be best to avoid bringing more of Gitoxide into the no‐Git‐backend build than necessary (it’s a concern for compile times too, after all), and that the explicit use of `git_{mailmap,actor}::*` would signal a distinction of these utility libraries from the `gix::*` used solely for the Git backend. However, if people would prefer to always bring in `gix` and just conditionalize features on it I’m happy to switch.

Byron replied:
I'd recommend to stick to plumbing as long as there is no disadvantage. With Mailmaps ,the involved thing is the initial setup in case you want to support all the git configuration around mailmaps as well. For that, it would be easiest to use `gix::Repository::*mailmap*()` to obtain a ready-made one.
*Note that I am probably missing a lot of context, as I only searched for `gix`, apologies if this is more noise than signal.*

emilazy replied:
We use the plumbing crate for interfacing with our Git backend, but in this case we’re using it purely for handling the `.mailmap` format directly, even when the Git backend is disabled, so there’s no Git repository lookup going on.

Byron replied:
I see, thanks for the clarification - in this case reading a single `.mailmap` with the plumbing crate will be fine.
And: yay to the modular crate approach :).

emilazy replied:
I’ve now dropped the unused `gix-date` dependency.

martinvonz reviewed (approved):
This looks great to me. Thanks!

Please give @yuja a chance to do another round of reviews too if he wants to.

yuja commented on cli/src/commit_templater.rs:
```diff
@@ -83,6 +85,8 @@ impl<'repo> CommitTemplateLanguage<'repo> {
                 .build_cache_extensions(&mut cache_extensions);
         }
 
+        let mailmap = Rc::new(get_current_mailmap(repo, workspace_id));
```
nit: It's probably better to cache parsed `mailmap` by caller, but this can be addressed later.

emilazy replied:
I looked at `WorkspaceCommandHelper` and it seems like it has methods that would change the contents of the working‐copy commit? So we’d need to keep track of invalidating that state correctly, which seems like it’s maybe not worth it for a total of ~5 calls per `jj log` command that don’t seem to show up in profiling. If it ends up being a performance issue then maybe a commit ID → parsed `Mailmap` cache would be the best solution?

yuja replied:
This kind of cache can be managed by `ReadonlyUserRepo`. I agree the cost of loading mailmap wouldn't matter, but doing that by caller will make error handling simpler.

emilazy replied:
I will take a look at that, thank you; I’m still getting used to the code organization.

emilazy replied:
Could you help me understand what kind of interface you’re envisioning here? I assume it’d go in `ReadonlyRepo` as user configuration isn’t relevant here, but I’m not sure what kind of state and methods would be expected. It seems like `ReadonlyRepo` doesn’t fix a specific working‐copy commit, so I guess it’d still need to take a workspace ID as a parameter? Are you thinking that `get_current_mailmap` would move entirely to a method on `ReadonlyRepo`? Does the revset querying code have access to a `ReadonlyRepo` too?

yuja replied:
Ah, I meant `ReadonlyUserRepo` in cli_util.rs, which can store `OnceCell<Mailmap>` as a cache.

I think it's also good to just move `get_current_mailmap()` to `WorkspaceCommandHelper` where we can propagate load/parsing error if needed.

emilazy replied:
Sorry for having trouble following the intent here. `ReadonlyUserRepo` doesn’t fix a specific workspace, right? So it’d have to be a `HashMap<WorkspaceId, Mailmap>` or something?

I’d definitely like to keep a helper in `jj_lib` for obtaining a `Mailmap` so that alternate frontends can easily do the right thing, per https://github.com/martinvonz/jj/pull/3951#discussion_r1653672458. But if we can figure out what the interface should look like here, it could presumably go directly in `jj_lib::repo::ReadOnlyRepo`, as the `id_prefix_context` in `ReadonlyUserRepo` isn’t relevant here and nothing about this would depend on user configuration.

yuja replied:
`ReadonlyUserRepo` is a private wrapper type for `WorkspaceCommandHelper` to manage cache lifecycle, and `WorkspaceCommandHelper` knows `WorkspaceId`.

So long as `Mailmap` is loaded from `@`, I don't think there is a good place to host it in `ReadonlyRepo`.

yuja commented on cli/tests/common/mod.rs:
```diff
@@ -160,7 +160,7 @@ impl TestEnvironment {
     }
 
-    fn get_ok(&self, mut cmd: assert_cmd::Command) -> (String, String) {
+    pub(crate) fn get_ok(&self, mut cmd: assert_cmd::Command) -> (String, String) {
```
nit: you can use `get_stdout_string(&assert)`, etc. instead. (see `test_no_user_configured()` for example.)

I'm not strongly against this change, but the function name is odd as you mentioned before.

emilazy replied:
I’ll look into this.

emilazy replied:
Okay, I’ve undone this change and just inlined what the test needs.

yuja commented on lib/src/default_index/revset_engine.rs:
```diff
@@ -1049,19 +1049,39 @@ fn build_predicate_fn(
+                let commit = store.get_commit(&entry.commit_id()).unwrap();
+                let author = mailmap.author(&commit);
+                pattern.matches(&author.name) || pattern.matches(&author.email)
+            })
```
Can't we transform the query by frontend?

```
author(foo)
-> author(foo)|author(exact:lookup_reverse(foo)[0])|author(exact:lookup_reverse(foo)[1])|..
```

It's not exactly the same (because the query also matches the original values), but I think it's practically okay. I think it's easier for revset engine backed by database to process query.

emilazy replied:
Right, I mentioned this in the commit message under “The mapping to canonical signatures is one‐way, and queries only match on the canonical form”. It seems a little tricky to translate back because the functions accept glob patterns; at the very least we would have to traverse the `Mailmap` structure and syntheize `author_raw(exact:…)` patterns for all possible matches. But I’m not entirely sure that would be correct in all cases, even if we accept the change in semantics. I’d have to think about it.

Rather than changing the semantics/implementation here, I think in a database backend, you could have a caching table like (commit ID or hash or something of `.mailmap`, raw signature that appears in commits, canonical signature from corresponding `.mailmap`), and an index for lookup? I don’t really know how those backends work since there isn’t a public example of one. Maybe you could even use a materialized view or something.

yuja replied:
I don't know the implementation either, and I guess Google backend can just ignore mailmap thingy at all (because .mailmap shouldn't exist there.)

That said, it'll be nice if mailmap handling is isolated to the frontend layer. Suppose the size of the mailmap is small compared to the commit history, I think it's okay to scan all entries to build a reverse mapping.

emilazy replied:
I had a think about this. If you wanted to avoid the additional table/view I mentioned, we can still obtain the same database performance benefits without changing semantics (which I’m a little reluctant to do, as it goes against the rubric of when raw signature content should affect UI I put in the commit message). `author(a*b*)` would optimize to `author(a*b*) & (author_raw(a*b*) | author_raw(exact:…) | author_raw(exact:…) | …)` based on the `.mailmap`. The latter can be queried on the database efficiently, and the `author(a*b*) &` is a quick post‐processing step on the results that ensures the semantics remain the same.

This wouldn’t be hard to implement, but I’m hesitant to add the complexity off the bat if it wouldn’t currently have any prospective users, especially as it would make performance strictly worse with the current backends. (Although considering the complaints I’ve heard from Googlers in the past about what a pain it is to get their username and name updated across all systems maybe they could do with having some kind of system here…)

yuja replied:
I don't think doing preprocessing is strictly worse in performance. If the query doesn't match any mailmap entry (which I think is common), there's no additional cost in later stage. OTOH, without preprocessing, we'll do binary search for each commit entry. I don't think the difference would be significant, though.

FWIW, if a database-backed revset engine transforms mailmap to table/view, they'll want a single mailmap resource than mailmap per expression node.

Anyway, I think this is a fundamental issue of mailmap design, and I just thought it would be nice if we can avoid spreading a mess to the backend layer.

emilazy replied:
The reason I say it would be strictly slower with the Git and local backends is that it would be increasing the number of comparisons. However, this probably isn’t really relevant. I take it that database backends don’t use `jj_lib::default_index::revset_engine` at all, then?

I’d like to discuss my semantics‐preserving solution further, if that’s alright. I think that desugaring/optimizing `author("new nam*")` to `author("new nam*") & (author_raw("new nam*") | author_raw(exact:"old name") | author_raw(exact:"old@email.example"))` would hopefully satisfy everyone here: it would provide all the same query performance advantages, not require any database schema adjustments, and be a very small burden on any implementors of alternate revset engines that want to support `.mailmap`. The idea is that there would be indexes on the raw signatures (as there are currently, presumably), and that the database’s optimized pattern query functions could be used on those, with the results then being post‐processed in either Rust code or the query language to ensure that they match the mapped name.

For instance, an index‐friendly query like the following could be issued, just like currently…

```sql
select …
from commits
where author_raw_name like "new nam%"
or author_raw_email like "new nam%"
or author_raw_name = "old name"
or author_raw_email = "old name"
or author_raw_name = "old@email.example"
or author_raw_email = "old@email.example"
```

…followed by a simple and quick `StringPattern::parse("new nam*").matches(mailmap.author(&commit))` filtering step in Rust to handle the `author("new nam*") &` condition. Obviously this is a common pattern when issuing database queries; I don’t know if every single revset function is translated to a direct database query filter with no Rust code currently, but if it’s desired for that property to be preserved then the post‐processing could also be implemented in the database query; any planner worth its name would do the fast index lookups before checking the additional condition, as with other queries that use complex functions. If there are already filter steps using Rust code, on the other hand, then this should “just work” with no additional logic or changes.

Anyway, I’m happy to implement this reverse‐lookup desugaring as requested, but if it’d slow things down on the in‐tree backends and Google doesn’t plan to use it, is it worth the additional logic right now? Maybe it’s just enough to know that we could do this without immediately implementing it?

emilazy replied:
Maybe it would be a good idea to make backends responsible for providing a `Mailmap` (which could be a trait instead?), and the Git and local backends could just share the current implementation for now. Again though I’m unsure how much value this would add if it wouldn’t matter to Google right now anyway.

emilazy replied:
Strawman design:

* `Mailmap` could become a trait.

* The `Backend` trait could gain e.g. `fn read_mailmap(&self, as_of_commit: &CommitId) -> impl Mailmap`. A database backend could stub this out if it doesn’t support the feature or ignore the parameter and return an interface to a database table. Handling the queries efficiently would be the responsibility of a backend that wants to support this, and could be handled as per the desugaring I outlined in https://github.com/martinvonz/jj/pull/3951#discussion_r1654369716.

* The in‐tree backends would share the current implementation.

Not sure if this is better or worse than the status quo, but I thought I should throw it out there.

yuja replied:
> Maybe it would be a good idea to make backends responsible for providing a `Mailmap` (which could be a trait instead?), and the Git and local backends could just share the current implementation for now.

Yeah, I think that's proper design for identity rewriting.

My understanding is that mailmap is ad-hoc solution and we decided to process mailmap at the frontend layer because of that? And, my point is that it's nice if mailmap handling does not cross the frontend-backend boudary. It might be a bit slow and inaccurate depending on the query, but I assume it's practically okay.

martinvonz replied:
> And, my point is that it's nice if mailmap handling does not cross the frontend-backend boudary.

I agree with this. At the same time, mailmap is a bit of a hack, so what Emily has currently implemented in this PR seems fine to me.

Another alternative might be to pass the `Mailmap` into `Index::evaluate_revset()`. What do you think about that?

yuja replied:
> Another alternative might be to pass the `Mailmap` into `Index::evaluate_revset()`.

That seems slightly better.

FWIW, I also consider mailmap as a hack, and that's the reason why I insist that the query result doesn't have to be 100% accurate. It's simpler to approximate mailmap query as union of matched mailmap entries than extending backend API.

PhilipMetzger replied:
> Strawman design:
> 
> * `Mailmap` could become a trait.
> * The `Backend` trait could gain e.g. `fn read_mailmap(&self, as_of_commit: &CommitId) -> impl Mailmap`. A database backend could stub this out if it doesn’t support the feature or ignore the parameter and return an interface to a database table. Handling the queries efficiently would be the responsibility of a backend that wants to support this, and could be handled as per the desugaring I outlined in [lib: add support for `.mailmap` files #3951 (comment)](https://github.com/martinvonz/jj/pull/3951#discussion_r1654369716).
> * The in‐tree backends would share the current implementation.
> 
> Not sure if this is better or worse than the status quo, but I thought I should throw it out there.

Yes, that'd be the approach if the pursue the identity work further, but clearly is unneeded complexity for this PR. As what the current `.mailmap` approach should be, there I heavily agree with Yuya and Martin.

emilazy replied:
> Another alternative might be to pass the `Mailmap` into `Index::evaluate_revset()`. What do you think about that?

I tried this, and it was a mess, sadly; the state needed to be piped through the whole evaluation system and it caused a lot of churn and other problems. It basically resulted in `ResolvedExpression`s not being fully resolved and having to patch that up in the evaluation layer, so in terms of these layering concerns it was a lot worse.

I wasn’t sure about how to move forward here, but then I realized that it is possible to implement the front‐end desugaring of the revset queries in a way that always gives correct `.mailmap` results and yields simple, efficient `RevsetExpression`s that should work great with database indexes. The backend now no longer needs to care about `.mailmap` support at all. Hopefully that will make everyone happy! I’m still finishing polishing it off, so I haven’t pushed it here yet, but I’ve opened https://github.com/martinvonz/jj/pull/3991 as an interim change I made along the way that resolves an old TODO and helped with the implementation.

(For what it’s worth: I do agree that `.mailmap` is essentially a pragmatic hack, an inherently incomplete measure, and a symptom of a deficient underlying design. Still, Git users are unlikely to ever get anything better, and for some people it’s a hack that is the difference between a marginally workable identity system and one that is completely unfit for purpose, so I do care about implementing it at least as well as Git does. I’m principled in my pragmatism; giving up on the principles of application I listed in the commit message results in concrete scenarios where it fails to meet the basic point of supporting `.mailmap`, in my view. Thankfully, it looks like it’s totally practical to address these layering concerns while still upholding them!)

yuja commented on lib/src/mailmap.rs:
```diff
@@ -0,0 +1,125 @@
+impl PartialEq for Mailmap {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.entries() == other.0.entries()
+    }
```
nit: might be better to use pointer-based comparison, and add TODO comment.
```// TODO: Refactor tests to not need the Eq trait```

`.entries()` constructs `Vec<_>`.

emilazy replied:
Would moving the tests over to `matches!` in another commit here work?

I’ve opened https://github.com/Byron/gitoxide/issues/1424 to ask for a better solution upstream; I might just send a PR so we can avoid the allocation here. Non‐semantic equality on immutable types makes me a bit nervous, personally.

yuja replied:
> Would moving the tests over to `matches!` in another commit here work?

I have draft patches to do that (moving to `assert_matches!` or `insta::assert_debug_snapshot`), so feel free to leave a TODO comment.

emilazy replied:
Done!

emilazy replied:
`gix-mailmap` 0.23.3 now has `Debug` and `Eq` implementations that don’t allocate, so I’ve switched to a `#[derive]` here. Adjusting the tests to not use equality comparison may still be a good idea, of course.

yuja commented on lib/src/mailmap.rs:
```diff
@@ -0,0 +1,125 @@
+/// treated the same way. Parse errors when reading the file are ignored, but
+/// the rest of the file will still be processed.
+pub fn get_current_mailmap(repo: &dyn Repo, workspace_id: &WorkspaceId) -> Mailmap {
+    let inner = || {
```
nit: can you add `TODO` comment about error handling?

emilazy replied:
Sure, but I’m not sure what it should say :) This is called by code that doesn’t really have a good way to bubble up errors further, and I don’t know how you’d handle them. I’m not entirely clear on what kinds of situations would result in failures here. I guess `// TODO: Figure out if any errors should be surfaced here.` should be enough?

yuja replied:
I/O-related error can occur under racy situation?, though it's likely to be detected earlier.

emilazy replied:
I added a comment.

emilazy commented:
I’ve pushed a pretty big rework of this; it feels cleaner now. Headline changes:

* Added a `Mailmap::empty` constructor to use instead of `Default::default` for clarity.

* Renamed `mailmap::{get → read}_current_mailmap` and added error reporting.

* Added a `WorkspaceCommandHelper::current_mailmap` convenience helper, although it currently only has one call site. (Maybe it should be `read_current_mailmap` too? Maybe it should keep the `async`?) Caching is a TODO for a later date, but performance impact seems negligible anyway.

* The details of handling `.mailmap` mappings are now dealt with entirely in the revset frontend; backends don’t have to care about it at all. The same correctness guarantees as the previous version are maintained, and the resulting expressions should be efficient to execute against database indexes. Everything is abstracted through simple functions like `RevsetExpression::author`; the mapping implementation is contained entirely within the private `RevsetExpression::signature_field` helper, which I have commented extensively to make the logic clear.

* Added even more paragraphs of detail to the main commit message!

I believe I’ve addressed all the previous review comments; hopefully this is acceptable to everyone now and we can get this merged! This turned into a bigger project than I expected when I picked it as my first public contribution to Jujutsu, but I’m grateful for all the feedback and I think this has ended up in a good place.

yuja commented on lib/src/revset.rs:
```diff
@@ -121,16 +121,22 @@ pub trait RevsetFilterExtension: std::fmt::Debug + Any {
+    /// Commits with author field matching the pattern.
+    Author(SignatureField, StringPattern),
+    /// Commits with committer field matching the pattern.
+    Committer(SignatureField, StringPattern),
```
Maybe these enum variants can be simply `AuthorName(_)`/`AuthorEmail(_)`/..?

We're going to introduce `AuthorDate(_)`, etc., so `Author<Thing>` seems more consistent. And we'll probably add `author_name()`/`_email()` revset function if needed.

yuja commented on lib/src/revset.rs:
```diff
@@ -2053,12 +2164,14 @@ impl<'a> RevsetParseContext<'a> {
         user_email: String,
         extensions: &'a RevsetExtensions,
         workspace: Option<RevsetWorkspaceContext<'a>>,
+        mailmap: Rc<Mailmap>,
```
nit: `Rc<_>` isn't needed, but templater depends on it. If `Mailmap` is cached by `WorkspaceCommandHelper`, we can replace `Rc` with `&'a`. That said, this isn't important, and can be addressed later. (I just find it looked a bit odd because the revset module doesn't require `Rc`.)

yuja commented on lib/src/revset.rs:
```diff
@@ -437,64 +438,162 @@ impl RevsetExpression {
+            self.clone()
+        } else {
+            RevsetExpression::minus(self, &RevsetExpression::union_all(others))
+        }
```
nit: I'll inline this to caller.

`x & ~none()` could be rewritten to `x` at later pass, but apparently we don't have a rule for that right now.

yuja commented on lib/src/revset.rs:
```diff
@@ -437,64 +438,162 @@ impl RevsetExpression {
+                old_field_value.map(|field_value| pattern_case_insensitive.matches(field_value));
+            let new_field_matches = pattern.matches(new_field_value);
+
+            if !new_field_matches && old_field_matches.unwrap_or(true) {
```
Doesn't `.unwrap_or(true)` mean we're building quite large negative sets? Mailmap entry  doesn't usually contain old name, I suppose. Even if it's more correct, I don't think the added cost is acceptable.

yuja replied:
FWIW, my original idea was to build query without `excludes`, meaning `author(x)` matches `x` in raw entry or mapped entry. This isn't perfect, but I considered it would be good enough. Mailmap is a bit of a hack.

If we need to guarantee that all shadowed matches are excluded, it's probably better to pass `Mailmap` to `evaluate_revset()` as a resource.

yuja commented on lib/src/revset.rs:
```diff
@@ -437,64 +438,162 @@ impl RevsetExpression {
+        //     }
+        //
+        // When matching raw signatures, the `Mailmap` is empty and this reduces to
+        // `raw_field(pattern)`.
```
nit: it's probably better to write unit tests for this query transformation function.

yuja commented on lib/src/str_util.rs:
```diff
@@ -174,6 +174,16 @@ impl StringPattern {
+            StringPattern::Exact(literal) => Cow::Owned(StringPattern::exact_i(literal)),
+            StringPattern::Substring(needle) => Cow::Owned(StringPattern::substring_i(needle)),
+            StringPattern::Glob(pattern) => Cow::Owned(StringPattern::GlobI(pattern.clone())),
+            _ => Cow::Borrowed(self),
```
nit: I'll list all variants exhaustively (to catch future bugs.)

yuja commented on lib/src/mailmap.rs:
```diff
@@ -0,0 +1,165 @@
+    pub fn new_name(&self) -> Option<&str> {
+        self.new_name.as_deref()
+    }
+}
```
nit: might be okay to simply make all fields public.

yuja reviewed (commented):
The direction of this PR looks good to me, thanks!

mati865 commented:
What is the status here? It'd be a shame to let it bit rot further.

emilazy commented:
Sorry for the inactivity; I was expecting this to be a relatively quick PR and burnt out a little on the review cycles. I still intend to rebase this and respond to review comments at some point, though I am not sure if it will be possible to fully satisfy everyone here.

lilyball commented:
I would really love to see this get merged as long as it's in a "good enough" state even if it's not perfect. There are repositories I don't want to use jujutsu with until it has mailmap support because I'd rather not see my deadname show up in the UI.

yuja commented:
> I'd rather not see my deadname show up in the UI.

FWIW, the display part (= templater + Commit object methods) is easy, and can be merged separately. IIRC, most of the discussion here is about `author()`/`committer()` revset.

emilazy commented:
(Accidentally hit the close button, whoops!)

I have local work on this that makes it a lot structurally cleaner and should address the design comments raised long, long ago and hopefully satisfy all parties. I still intend to pick this back up and get it merged at some point, it’s just a fair amount of work to get it completely ready and I want to make sure it’s polished enough to not have to go through another bunch of review cycles :)

maddiemort commented:
Just FYI, today I finished rebasing this work onto the 0.28.2 release commit. I don't think I did it perfectly, and it's certainly not up to the kind of quality standard where I'd submit a PR for it, but for anyone who's been wanting to update to a newer version of jj, hopefully this will tide you over 😄

https://github.com/maddiemort/jj/tree/updated-mailmap

(for me, this was a jump all the way from 0.20 to 0.28!)

emilazy commented:
Thanks! I’m sorry I’ve neglected this PR for so long – I hope I’ll have time over the next couple months to rebase and push out my rewritten version that should have a much better chance of landing in trunk.

maddiemort commented:
Don't worry at all! When you do come back to this, you may find the work on my branch helpful - the four commits that retain you as author don't contain changes other than your work plus conflict resolution, so you might be able to pick them up wholesale for your own rebase.
